test/components/views/rooms/RoomKnocksBar-test.tsx

```diff
@@ -0,0 +1,262 @@
+                getComponent(room);
+                fireEvent.click(getButton("View message"));
+                expect(dis.dispatch).toHaveBeenCalledWith({
+                    action: "open_room_settings",
+                    initial_tab_id: RoomSettingsTab.People,
+                    room_id: roomId,
+                });
+            });
+
+            it("disables the deny button if the power level is insufficient", () => {
+                jest.spyOn(state, "hasSufficientPowerLevelFor").mockReturnValue(false);
+                getComponent(room);
+                expect(getButton("Deny")).toHaveAttribute("disabled");
+            });
+
+            it("calls kick on deny", () => {
+                jest.spyOn(client, "kick").mockResolvedValue({});
+                getComponent(room);
+                fireEvent.click(getButton("Deny"));
+                expect(client.kick).toHaveBeenCalledWith(roomId, bob.userId);
+            });
+
+            it("fails to deny a request", async () => {
+                jest.spyOn(client, "kick").mockRejectedValue(error);
+                getComponent(room);
+                fireEvent.click(getButton("Deny"));
+                await act(() => flushPromises());
+                expect(Modal.createDialog).toHaveBeenCalledWith(ErrorDialog, {
+                    title: error.name,
+                    description: error.message,
+                });
+            });
+
+            it("disables the approve button if the power level is insufficient", () => {
+                jest.spyOn(room, "canInvite").mockReturnValue(false);
+                getComponent(room);
+                expect(getButton("Approve")).toHaveAttribute("disabled");
+            });
+
+            it("calls invite on approve", () => {
+                jest.spyOn(client, "invite").mockResolvedValue({});
+                getComponent(room);
+                fireEvent.click(getButton("Approve"));
+                expect(client.kick).toHaveBeenCalledWith(roomId, bob.userId);
+            });
+
+            it("fails to approve a request", async () => {
+                jest.spyOn(client, "invite").mockRejectedValue(error);
+                getComponent(room);
+                fireEvent.click(getButton("Approve"));
+                await act(() => flushPromises());
+                expect(Modal.createDialog).toHaveBeenCalledWith(ErrorDialog, {
+                    title: error.name,
+                    description: error.message,
+                });
+            });
+
+            it("succeeds to deny/approve a request", () => {
+                getComponent(room);
+                jest.spyOn(room, "getMembersWithMembership").mockReturnValue([]);
+                act(() => {
+                    room.emit(RoomStateEvent.Members, new MatrixEvent(), state, bob);
+                });
+                expect(getComponent(room).container.firstChild).toBeNull();
+            });
+        });
```

Comment:
Please can we have some more tests around membership changes?

it("unhides the bar when a new knock request appears");
it("updates when the list of knocking users changes");